coffee/cse.py

```diff
@@ -0,0 +1,407 @@
+
+                # Calculate the operation count for /t/ if we applied expansion + fact
+                reads = []
+                for read, cost in t.reads_costs.items():
+                    if read.urepr in new_trace:
+                        reads.extend(new_trace[read.urepr].reads or [read.urepr])
+                        t_outloop_cost += new_trace[read.urepr].project*cost
+                    else:
+                        reads.extend([read.urepr])
+
+                # Factorization will kill duplicates and increase the number of sums
+                # in the outer loop
+                fact_syms = {s.urepr if isinstance(s, Symbol) else s for s in reads}
+                t_outloop_cost += len(reads) - len(fact_syms)
+
+                # Note: if n=len(fact_syms), then we'll have n prods, n-1 sums
+                t_inloop_cost += 2*len(fact_syms) - 1
+
+                # Add to the total and scale up by the corresponding number of iterations
+                total_outloop_cost += t_outloop_cost*t.niters_after_licm
+                level_inloop_cost += t_inloop_cost*t.niters
+
+                # Update the trace because we want to track the cost after "pushing" the
+                # temporaries on which /t/ depends into /t/ itself
+                new_trace[t.urepr].reads_costs = {s: 1 for s in fact_syms}
+
+            # Some temporaries at levels < /i/ may also appear in:
+            # 1) subsequent loops
+            # 2) levels beyond /i/
+            for t in list(flatten([levels[j] for j in range(level)])):
+                if any(ir.urepr not in new_trace for ir in t.is_read) or \
+                        any(new_trace[ir.urepr].level > level for ir in t.is_read):
+                    # Note: condition 1) is basically saying "if I'm read from
+                    # a temporary that is not in this loop's trace, then I must
+                    # be read in some other loops".
+                    level_inloop_cost += t.flops*t.niters
+
+            # Total cost = cost_after_fact_up_to_level + cost_inloop_cse
+            #            = cost_hoisted_subexprs + cost_inloop_fact + cost_inloop_cse
+            uptolevel_cost = cse_cost + total_outloop_cost + level_inloop_cost
+            uptolevel_cost += self._cost_cse(fact_levels, (level + 1, bounds[1]))
+
+            # Update the best alternative
+            if uptolevel_cost < best[2]:
+                best = (bounds[0], level, uptolevel_cost)
+
+            cse = self._cost_cse(fact_levels, (level + 1, bounds[1]))
+
+        if verbose:
+            print BLUE % ("Cost model :: unpicking CSE between levels [%d, %d]:" % bounds),
+            print BLUE % ("cost=%d (cse=%d, outloop=%d, inloop_fact=%d, inloop_cse=%d)" %
+                          (uptolevel_cost, cse_cost, total_outloop_cost,
+                           level_inloop_cost, cse))
+
+        return best
+
+    def unpick(self):
+        fors = visit(self.header, info_items=['fors'])['fors']
```

Comment:
just the main cseunpicker function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)